src/stringify.js

```diff
@@ -0,0 +1,60 @@
+import validate from './validate.js';
+
+/**
+ * Convert array of 16 byte values to UUID string format of the form:
+ * XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX
+ */
+const byteToHex = [];
+
+for (let i = 0; i < 256; ++i) {
+  byteToHex.push((i + 0x100).toString(16).substr(1));
+}
+
+function stringify(arr, offset = 0) {
+  // Note: Be careful editing this code!  It's been tuned for performance
+  // and works in ways you may not expect. See https://github.com/uuidjs/uuid/pull/434
+  const uuid = (
+    byteToHex[arr[offset + 0]] +
+    byteToHex[arr[offset + 1]] +
+    byteToHex[arr[offset + 2]] +
+    byteToHex[arr[offset + 3]] +
+    '-' +
+    byteToHex[arr[offset + 4]] +
+    byteToHex[arr[offset + 5]] +
+    '-' +
+    byteToHex[arr[offset + 6]] +
+    byteToHex[arr[offset + 7]] +
+    '-' +
+    byteToHex[arr[offset + 8]] +
+    byteToHex[arr[offset + 9]] +
+    '-' +
+    byteToHex[arr[offset + 10]] +
+    byteToHex[arr[offset + 11]] +
+    byteToHex[arr[offset + 12]] +
+    byteToHex[arr[offset + 13]] +
+    byteToHex[arr[offset + 14]] +
+    byteToHex[arr[offset + 15]]
+  ).toLowerCase();
+
+  // Sanity check for valid UUID.  This works because if any of
+  // the input array values don't map to a defined hex octet, the string length
+  // will get blown out (e.g. "74af23d8-85undefined2c44-...")
+  //
+  // This is a somewhat crude check, but avoids having to check each value
+  // individually.
+  if (uuid.length !== 36) {
+    throw new TypeError(
+      'Stringified UUID is invalid (All input values must be integers between 0 and 255)',
+    );
+  }
```

Comment:
If we do the expensive validation afterwards this would only be a perf improvement for the case where people try to stringify malformed byte arrays. I don't think this is worth the extra code.

Maybe just keep the additional validation?

What's the perf impact of the additional validation on `v4` and `v1` UUID generation?

Reply:
Good catch.  Making that change.

Regarding perf, validating in stringify does have an impact on uuid generation but, personally, not enough that I'm overly concerned.  I suppose we could add a `noValidate` arg to `stringify()` to allow callers to suppress the check, but I'm okay leaving that out for the time being and waiting until someone makes a compelling case for it.

**Perf with stringify() validation**:

```
uuidv1() x 1,613,447 ops/sec ±0.60% (94 runs sampled)
uuidv1() fill existing array x 9,143,504 ops/sec ±0.34% (99 runs sampled)
uuidv4() x 397,873 ops/sec ±1.02% (89 runs sampled)
uuidv4() fill existing array x 556,775 ops/sec ±0.59% (97 runs sampled)
uuidv3() x 212,358 ops/sec ±1.71% (88 runs sampled)
uuidv5() x 219,256 ops/sec ±1.88% (88 runs sampled)
```

**Perf w/out stringify() validation**:

```
uuidv1() x 2,183,327 ops/sec ±0.59% (90 runs sampled)
uuidv1() fill existing array x 9,118,317 ops/sec ±0.21% (93 runs sampled)
uuidv4() x 537,945 ops/sec ±0.36% (93 runs sampled)
uuidv4() fill existing array x 554,379 ops/sec ±0.37% (95 runs sampled)
uuidv3() x 257,289 ops/sec ±0.56% (83 runs sampled)
uuidv5() x 254,984 ops/sec ±2.09% (76 runs sampled)
```